llvm/tools/llvm-foreach/llvm-foreach.cpp

```diff
@@ -0,0 +1,207 @@
+    std::string Postfix = Arg.substr(Arg.find_last_of(Replace) + 1);
+    return {ArgNum, Prefix, Postfix};
+  };
+
+  // Find args to replace with filenames from input list.
+  std::vector<ArgumentReplace> InReplaceArgs;
+  ArgumentReplace OutReplaceArg;
+  for (size_t i = 0; i < Args.size(); ++i) {
+    for (auto &Replace : Replaces)
+      if (Args[i].contains(Replace)) {
+        InReplaceArgs.push_back(CreateArgumentReplace(i, Args[i], Replace));
+      }
+
+    if (!OutReplace.empty() && Args[i].contains(OutReplace))
+      OutReplaceArg = CreateArgumentReplace(i, Args[i], OutReplace);
+  }
+
+  // Emit an error if user requested replace output file in the command but
+  // replace string is not found.
+  if (!OutReplace.empty() && OutReplaceArg.ArgNum < 0)
+    error("Couldn't find replace string for output in the command.");
+
+  // Make sure that specified program exists, emit an error if not.
+  std::string Prog =
+      ExitOnErr(errorOrToExpected(sys::findProgramByName(Args[0])));
+
+  std::vector<std::vector<std::string>> FileLists(LineIterators.size());
+  int PrevNumOfLines = 0;
+  for (size_t i = 0; i < FileLists.size(); ++i) {
+    int NumOfLines = 0;
+    for (; !LineIterators[i].is_at_eof(); ++LineIterators[i]) {
+      FileLists[i].push_back(LineIterators[i]->str());
+      NumOfLines++;
+    }
+    if (i != 0 && NumOfLines != PrevNumOfLines)
+      error("All input file lists must have same number of lines!");
+    PrevNumOfLines = NumOfLines;
+  }
+
+  std::error_code EC;
+  std::string ResOutArg;
+  std::vector<std::string> ResInArgs(InReplaceArgs.size());
+  std::string ResFileList = "";
+  for (size_t j = 0; j != FileLists[0].size(); ++j) {
+    for (size_t i = 0; i < InReplaceArgs.size(); ++i) {
+      ArgumentReplace CurReplace = InReplaceArgs[i];
+      ResInArgs[i] = (Twine(CurReplace.Prefix) + Twine(FileLists[i][j]) +
+                      Twine(CurReplace.Postfix))
+                         .str();
+      Args[CurReplace.ArgNum] = ResInArgs[i];
+    }
+
+    SmallString<128> Path;
+    if (!OutReplace.empty()) {
+      // Create a file for command result. Add file name to output
+      // file list if needed.
+      std::string TempFileNameBase = sys::path::stem(OutReplace);
+      if (OutDirectory.empty())
+        EC = sys::fs::createTemporaryFile(TempFileNameBase, OutFilesExt, Path);
+      else {
+        SmallString<128> PathPrefix(OutDirectory);
+        llvm::sys::path::append(PathPrefix,
+                                TempFileNameBase + "-%%%%%%." + OutFilesExt);
+        EC = sys::fs::createUniqueFile(PathPrefix, Path);
+      }
+      error(EC, "Could not create a file for command output.");
+
+      ResOutArg = (Twine(OutReplaceArg.Prefix) + Twine(Path) +
+                   Twine(OutReplaceArg.Postfix))
+                      .str();
+      Args[OutReplaceArg.ArgNum] = ResOutArg;
+
+      if (!OutputFileList.empty())
+        ResFileList = (Twine(ResFileList) + Twine(Path) + Twine("\n")).str();
```

Comment:
There is no need to keep concatenating file names into a string, you can just write them to `OutputFileList` file.

Reply:
Okay, done.